aggregator/src/aggregator/aggregation_job_driver.rs

```diff
@@ -678,41 +678,56 @@ impl AggregationJobDriver {
         let report_aggregations_to_write = Arc::new(report_aggregations_to_write);
-        let aggregation_job_to_write = Arc::new(aggregation_job_to_write);
         let accumulator = Arc::new(accumulator);
         datastore
             .run_tx_with_name("step_aggregation_job_2", |tx| {
                 let (
                     vdaf,
-                    report_aggregations_to_write,
                     aggregation_job_to_write,
+                    report_aggregations_to_write,
                     accumulator,
                     lease,
                 ) = (
                     Arc::clone(&vdaf),
-                    Arc::clone(&report_aggregations_to_write),
                     Arc::clone(&aggregation_job_to_write),
+                    Arc::clone(&report_aggregations_to_write),
                     Arc::clone(&accumulator),
                     Arc::clone(&lease),
                 );
                 Box::pin(async move {
-                    let report_aggregations_future =
-                        try_join_all(report_aggregations_to_write.iter().map(
-                            |report_aggregation| tx.update_report_aggregation(report_aggregation),
-                        ));
-                    let aggregation_job_future =
-                        tx.update_aggregation_job(&aggregation_job_to_write);
-                    let batch_aggregations_future = accumulator.flush_to_datastore(tx, &vdaf);
+                    let unwritable_reports = accumulator.flush_to_datastore(tx, &vdaf).await?;
 
                     try_join!(
                         tx.release_aggregation_job(&lease),
-                        report_aggregations_future,
-                        aggregation_job_future,
-                        batch_aggregations_future,
+                        tx.update_aggregation_job(&aggregation_job_to_write),
+                        try_join_all(report_aggregations_to_write.iter().map(|ra| {
+                            let report_is_unwritable = unwritable_reports.contains(ra.report_id());
+                            async move {
+                                if report_is_unwritable {
+                                    tx.update_report_aggregation(
+                                        &ra.clone()
+                                            .with_state(ReportAggregationState::Failed(
+                                                ReportShareError::BatchCollected,
+                                            ))
+                                            .with_last_prep_step(Some(PrepareStep::new(
+                                                *ra.report_id(),
+                                                PrepareStepResult::Failed(
+                                                    ReportShareError::BatchCollected,
+                                                ),
+                                            ))),
+                                    )
+                                    .await
```

Comment:
Similar observation as in the helper/`aggregation_job_continue.rs` case: should we update the aggregation job state to finished if all reports are failed with `BatchCollected`?

Reply:
Good catch, in both locations.

The helper implementation was fixable by just moving some code around; fixing things here required slightly more effort due to needing to modify things which must be in an `Arc` since they are passed into a transaction handler. The most complicated bit was using a `Cow` to avoid copying report aggregations in the (overwhelmingly common) case that there is no race condition.